README.md
# Http Component
![version](https://img.shields.io/badge/version-1.1.0-brightgreen.svg?style=flat-square "Version")
[![MIT License](https://img.shields.io/badge/license-MIT-blue.svg?style=flat-square)](https://github.com/flextype-components/http/blob/master/LICENSE)

Simple Http Component to work with Http request, response and urls.

### Installation

```
composer require flextype-components/http
```

### Usage

Set header status
```php
Http::setResponseStatus(404);
```

Redirects the browser to a page specified by the url argument.
```php
Http::redirect('http://flextype.org');
```

Set one or multiple request headers.
```php
Http::setRequestHeaders('Location: http://flextype.org');
```

Get
```php
$action = Http::get('action');
```

Post
```php
$username = Http::post('username');
```

Returns whether this is an ajax request or not
```php
if (Http::isAjaxRequest()) {
  // do something...
}
```

Terminate request
```php
Http::requestShutdown();
```


Gets the base URL
```php
echo Http::getBaseUrl();
```

Gets current URL
```php
echo Http::getCurrentUrl();
```

Get Uri String
```php
$uri_string = Http::getUriString();
```

Get Uri Segments
```php
$uri_segments = Http::getUriSegments();
```

Get Uri Segment
```php
$uri_segment = Http::getUriSegment(1);
```

## License
See [LICENSE](https://github.com/flextype-components/http/blob/master/LICENSE)


CHANGELOG.md
# v1.1.0, 2018-04-07
* PHP7 Support Added
* New Http class added instead of Url, Request and Response classes

# v1.0.2, 2015-10-27
* Added Composer Support
* Documentation updates

# v1.0.1, 2015-10-17
* MIT LICENSE instead of GNU GPL v3

# v1.0.0, 2015-10-07
* Initial release


Request.php
<?php

namespace Flextype\Component\Http\Request;

use Flextype\Component\Arr\Arr;

/**
 * This file is part of the Force Components.
 *
 * (c) Romanenko Sergey / Awilum <awilum@msn.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */


class Request
{
    /**
     * Protected constructor since this is a static class.
     *
     * @access  protected
     */
    protected function __construct()
    {
        // Nothing here
    }

    /**
     * Redirects the browser to a page specified by the $url argument.
     *
     *  <code>
     *		Request::redirect('test');
     *  </code>
     *
     * @param string  $url    The URL
     * @param integer $status Status
     * @param integer $delay  Delay
     */
    public static function redirect($url, $status = 302, $delay = null)
    {
        // Redefine vars
        $url    = (string) $url;
        $status = (int) $status;

        // Status codes
        $messages = array();
        $messages[301] = '301 Moved Permanently';
        $messages[302] = '302 Found';

        // Is Headers sent ?
        if (headers_sent()) {
            echo "<script>document.location.href='" . $url . "';</script>\n";
        } else {

            // Redirect headers
            Request::setHeaders('HTTP/1.1 ' . $status . ' ' . Arr::get($messages, $status, 302));

            // Delay execution
            if ($delay !== null) {
                sleep((int) $delay);
            }

            // Redirect
            Request::setHeaders("Location: $url");

            // Shutdown request
            Request::shutdown();
        }
    }

    /**
     * Set one or multiple headers.
     *
     *  <code>
     *		Request::setHeaders('Location: http://site.com/');
     *  </code>
     *
     * @param mixed $headers String or array with headers to send.
     */
    public static function setHeaders($headers)
    {
        // Loop elements
        foreach ((array) $headers as $header) {

            // Set header
            header((string) $header);
        }
    }

    /**
     * Get
     *
     *  <code>
     *		$action = Request::get('action');
     *  </code>
     *
     * @param string $key Key
     * @param mixed
     */
    public static function get($key)
    {
        return Arr::get($_GET, $key);
    }

    /**
     * Post
     *
     *  <code>
     *		$login = Request::post('login');
     *  </code>
     *
     * @param string $key Key
     * @param mixed
     */
    public static function post($key)
    {
        return Arr::get($_POST, $key);
    }

    /**
     * Returns whether this is an ajax request or not
     *
     *  <code>
     *		if (Request::isAjax()) {
     *			// do something...
     *		}
     *  </code>
     *
     * @return boolean
     */
    public static function isAjax()
    {
        return isset($_SERVER['HTTP_X_REQUESTED_WITH']) && $_SERVER['HTTP_X_REQUESTED_WITH'] === 'XMLHttpRequest';
    }

    /**
     * Terminate request
     *
     *  <code>
     *		Request::shutdown();
     *  </code>
     *
     */
    public static function shutdown()
    {
        exit(0);
    }
}


Response.php
<?php

namespace Flextype\Component\Http\Response;

/**
 * This file is part of the Force Components.
 *
 * (c) Romanenko Sergey / Awilum <awilum@msn.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */


class Response
{
    /**
     * HTTP status codes and messages
     *
     * @var array
     */
    public static $messages = array(
        100 => 'Continue',
        101 => 'Switching Protocols',
        102 => 'Processing',            // RFC2518
        200 => 'OK',
        201 => 'Created',
        202 => 'Accepted',
        203 => 'Non-Authoritative Information',
        204 => 'No Content',
        205 => 'Reset Content',
        206 => 'Partial Content',
        207 => 'Multi-Status',          // RFC4918
        208 => 'Already Reported',      // RFC5842
        226 => 'IM Used',               // RFC3229
        300 => 'Multiple Choices',
        301 => 'Moved Permanently',
        302 => 'Found',
        303 => 'See Other',
        304 => 'Not Modified',
        305 => 'Use Proxy',
        306 => 'Reserved',
        307 => 'Temporary Redirect',
        308 => 'Permanent Redirect',    // RFC-reschke-http-status-308-07
        400 => 'Bad Request',
        401 => 'Unauthorized',
        402 => 'Payment Required',
        403 => 'Forbidden',
        404 => 'Not Found',
        405 => 'Method Not Allowed',
        406 => 'Not Acceptable',
        407 => 'Proxy Authentication Required',
        408 => 'Request Timeout',
        409 => 'Conflict',
        410 => 'Gone',
        411 => 'Length Required',
        412 => 'Precondition Failed',
        413 => 'Request Entity Too Large',
        414 => 'Request-URI Too Long',
        415 => 'Unsupported Media Type',
        416 => 'Requested Range Not Satisfiable',
        417 => 'Expectation Failed',
        418 => 'I\'m a teapot',                                               // RFC2324
        422 => 'Unprocessable Entity',                                        // RFC4918
        423 => 'Locked',                                                      // RFC4918
        424 => 'Failed Dependency',                                           // RFC4918
        425 => 'Reserved for WebDAV advanced collections expired proposal',   // RFC2817
        426 => 'Upgrade Required',                                            // RFC2817
        428 => 'Precondition Required',                                       // RFC6585
        429 => 'Too Many Requests',                                           // RFC6585
        431 => 'Request Header Fields Too Large',                             // RFC6585
        500 => 'Internal Server Error',
        501 => 'Not Implemented',
        502 => 'Bad Gateway',
        503 => 'Service Unavailable',
        504 => 'Gateway Timeout',
        505 => 'HTTP Version Not Supported',
        506 => 'Variant Also Negotiates (Experimental)',                      // RFC2295
        507 => 'Insufficient Storage',                                        // RFC4918
        508 => 'Loop Detected',                                               // RFC5842
        510 => 'Not Extended',                                                // RFC2774
        511 => 'Network Authentication Required',                             // RFC6585
    );

    /**
     * Protected constructor since this is a static class.
     *
     * @access  protected
     */
    protected function __construct()
    {
        // Nothing here
    }

    /**
     * Set header status
     *
     *  <code>
     *		Response::status(404);
     *  </code>
     *
     * @param integer $status Status code
     */
    public static function status($status)
    {
        if (array_key_exists($status, Response::$messages)) {
            header('HTTP/1.1 ' . $status . ' ' . Response::$messages[$status]);
        }
    }
}


Http.php
<?php

/**
 * @package Flextype Components
 *
 * @author Sergey Romanenko <awilum@yandex.ru>
 * @link http://components.flextype.org
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Flextype\Component\Http;

use Flextype\Component\Arr\Arr;

class Http
{
    /**
     * HTTP status codes and messages
     *
     * @var array
     */
    public static $http_status_messages = [
        100 => 'Continue',
        101 => 'Switching Protocols',
        102 => 'Processing',            // RFC2518
        200 => 'OK',
        201 => 'Created',
        202 => 'Accepted',
        203 => 'Non-Authoritative Information',
        204 => 'No Content',
        205 => 'Reset Content',
        206 => 'Partial Content',
        207 => 'Multi-Status',          // RFC4918
        208 => 'Already Reported',      // RFC5842
        226 => 'IM Used',               // RFC3229
        300 => 'Multiple Choices',
        301 => 'Moved Permanently',
        302 => 'Found',
        303 => 'See Other',
        304 => 'Not Modified',
        305 => 'Use Proxy',
        306 => 'Reserved',
        307 => 'Temporary Redirect',
        308 => 'Permanent Redirect',    // RFC-reschke-http-status-308-07
        400 => 'Bad Request',
        401 => 'Unauthorized',
        402 => 'Payment Required',
        403 => 'Forbidden',
        404 => 'Not Found',
        405 => 'Method Not Allowed',
        406 => 'Not Acceptable',
        407 => 'Proxy Authentication Required',
        408 => 'Request Timeout',
        409 => 'Conflict',
        410 => 'Gone',
        411 => 'Length Required',
        412 => 'Precondition Failed',
        413 => 'Request Entity Too Large',
        414 => 'Request-URI Too Long',
        415 => 'Unsupported Media Type',
        416 => 'Requested Range Not Satisfiable',
        417 => 'Expectation Failed',
        418 => 'I\'m a teapot',                                               // RFC2324
        422 => 'Unprocessable Entity',                                        // RFC4918
        423 => 'Locked',                                                      // RFC4918
        424 => 'Failed Dependency',                                           // RFC4918
        425 => 'Reserved for WebDAV advanced collections expired proposal',   // RFC2817
        426 => 'Upgrade Required',                                            // RFC2817
        428 => 'Precondition Required',                                       // RFC6585
        429 => 'Too Many Requests',                                           // RFC6585
        431 => 'Request Header Fields Too Large',                             // RFC6585
        500 => 'Internal Server Error',
        501 => 'Not Implemented',
        502 => 'Bad Gateway',
        503 => 'Service Unavailable',
        504 => 'Gateway Timeout',
        505 => 'HTTP Version Not Supported',
        506 => 'Variant Also Negotiates (Experimental)',                      // RFC2295
        507 => 'Insufficient Storage',                                        // RFC4918
        508 => 'Loop Detected',                                               // RFC5842
        510 => 'Not Extended',                                                // RFC2774
        511 => 'Network Authentication Required',                             // RFC6585
    ];

    /**
     * Set response header status
     *
     * Http::setResponseStatus(404);
     *
     * @param integer $status Status code
     * @return void
     */
    public static function setResponseStatus(int $status) : void
    {
        if (array_key_exists($status, Http::$http_status_messages)) {
            header('HTTP/1.1 ' . $status . ' ' . Http::$http_status_messages[$status]);
        }
    }

    /**
     * Redirects the browser to a page specified by the $url argument.
     *
     * Http::redirect('test');
     *
     * @param string  $url    The URL
     * @param integer $status Status
     * @param integer $delay  Delay
     */
    public static function redirect(string $url, int $status = 302, int $delay = null)
    {
        // Status codes
        $messages = [];
        $messages[301] = '301 Moved Permanently';
        $messages[302] = '302 Found';

        // Is Headers sent ?
        if (headers_sent()) {
            echo "<script>document.location.href='" . $url . "';</script>\n";
        } else {

            // Redirect headers
            Http::setRequestHeaders('HTTP/1.1 ' . $status . ' ' . Arr::get($messages, $status, 302));

            // Delay execution
            if ($delay !== null) {
                sleep((int) $delay);
            }

            // Redirect
            Http::setRequestHeaders("Location: $url");

            // Shutdown request
            Http::requestShutdown();
        }
    }

    /**
     * Set one or multiple headers.
     *
     * Http::setRequestHeaders('Location: http://site.com/');
     *
     * @param mixed $headers String or array with headers to send.
     */
    public static function setRequestHeaders($headers)
    {
        // Loop elements
        foreach ((array) $headers as $header) {

            // Set header
            header((string) $header);
        }
    }

    /**
     * Get
     *
     * $action = Http::get('action');
     *
     * @param string $key Key
     * @param mixed
     */
    public static function get(string $key)
    {
        return Arr::get($_GET, $key);
    }

    /**
     * Post
     *
     * $login = Http::post('login');
     *
     * @param string $key Key
     * @param mixed
     */
    public static function post(string $key)
    {
        return Arr::get($_POST, $key);
    }

    /**
     * Gets the base URL
     *
     * echo Http::getBaseUrl();
     *
     * @return string
     */
    public static function getBaseUrl() : string
    {
        $https = (isset($_SERVER['HTTPS']) && strtolower($_SERVER['HTTPS']) == 'on') ? 'https://' : 'http://';

        return $https . rtrim(rtrim($_SERVER['HTTP_HOST'], '\\/') . dirname($_SERVER['PHP_SELF']), '\\/');
    }

    /**
     * Gets current URL
     *
     * echo Http::getCurrentUrl();
     *
     * @return string
     */
    public static function getCurrentUrl() : string
    {
        return (!empty($_SERVER['HTTPS'])) ? "https://".$_SERVER['SERVER_NAME'].$_SERVER['REQUEST_URI'] : "http://".$_SERVER['SERVER_NAME'].$_SERVER['REQUEST_URI'];
    }

    /**
     * Get Uri String
     *
     * $uri_string = Http::getUriString();
     *
     * @access  public
     * @return string
     */
    public static function getUriString() : string
    {
        // Get request url and script url
        $url = '';
        $request_url = (isset($_SERVER['REQUEST_URI'])) ? $_SERVER['REQUEST_URI'] : '';
        $script_url  = (isset($_SERVER['PHP_SELF'])) ? $_SERVER['PHP_SELF'] : '';

        // Get our url path and trim the / of the left and the right
        if ($request_url != $script_url) {
            $url = trim(preg_replace('/'. str_replace('/', '\/', str_replace('index.php', '', $script_url)) .'/', '', $request_url, 1), '/');
        }

        $url = preg_replace('/\?.*/', '', $url); // Strip query string

        return $url;
    }

    /**
     * Get Uri Segments
     *
     * $uri_segments = Http::getUriSegments();
     *
     * @access  public
     * @return array
     */
    public static function getUriSegments() : array
    {
        return explode('/', self::getUriString());
    }

    /**
     * Get Uri Segment
     *
     * $uri_segment = Http::getUriSegment(1);
     *
     * @access  public
     * @param int $segment segment
     * @return string
     */
    public static function getUriSegment(int $segment) : string
    {
        $segments = self::getUriSegments();
        return isset($segments[$segment]) ? $segments[$segment] : null;
    }

    /**
     * Returns whether this is an ajax request or not
     *
     * if (Http::isAjaxRequest()) {
     *   // do something...
     * }
     *
     * @return boolean
     */
    public static function isAjaxRequest() : bool
    {
        return isset($_SERVER['HTTP_X_REQUESTED_WITH']) && $_SERVER['HTTP_X_REQUESTED_WITH'] === 'XMLHttpRequest';
    }

    /**
     * Terminate request
     *
     *  Http::requestShutdown();
     *
     */
    public static function requestShutdown()
    {
        exit(0);
    }

}
